src/js/profiling/integration.ts

```diff
@@ -0,0 +1,190 @@
+import type { Envelope, Event, EventProcessor, Hub, Integration, Profile, Transaction  } from '@sentry/types';
+import { logger, uuid4 } from '@sentry/utils';
+
+import { isHermesEnabled } from '../utils/environment';
+import { PROFILE_QUEUE } from './cache';
+import { startProfiling, stopProfiling } from './hermes';
+import { addProfilesToEnvelope, createProfilingEvent, findProfiledTransactionsFromEnvelope } from './utils';
+
+export const MAX_PROFILE_DURATION_MS = 30 * 1e6;
+
+/**
+ * Profiling integration creates a profile for each transaction and adds it to the event envelope.
+ *
+ * @experimental
+ */
+export class HermesProfiling implements Integration {
+  /**
+   * @inheritDoc
+   */
+  public static id: string = 'HermesProfiling';
+
+  /**
+   * @inheritDoc
+   */
+  public name: string = HermesProfiling.id;
+
+  private _getCurrentHub?: () => Hub;
+
+  private _currentProfile:
+    | {
+        profile_id: string;
+        startTimestampNs: number;
+      }
+    | undefined;
+
+  private _currentProfileTimeout: number | undefined;
+
+  /**
+   * @inheritDoc
+   */
+  public setupOnce(_: (e: EventProcessor) => void, getCurrentHub: () => Hub): void {
+    if (!isHermesEnabled()) {
+      logger.log('[Profiling] Hermes is not enabled, not adding profiling integration.');
+      return;
+    }
+
+    this._getCurrentHub = getCurrentHub;
+    const client = getCurrentHub().getClient();
+
+    if (!client || typeof client.on !== 'function') {
+      return;
+    }
+
+    client.on('startTransaction', (transaction: Transaction) => {
+      typeof this._currentProfileTimeout === 'number' && clearTimeout(this._currentProfileTimeout);
+      this._finishCurrentProfile();
+
+      const shouldStartProfiling = this._shouldStartProfiling(transaction);
+      if (!shouldStartProfiling) {
+        return;
+      }
+
+      this._startNewProfile(transaction);
+      setTimeout(
```

Comment:
I dont think you are ever setting this._currentProfileTimeout?

Reply:
Thank you, that's a good catch. I will also add a test for it.